### ISSUE TITLE
System.Data.SqlClient 4.8.2 throws TransactionAbortedException (the connection has been broken)

### ISSUE THREAD

**Issue by MichaelLogutov** [dotnet/SqlClient#729]
### Description

After updating System.Data.SqlClient version from old 4.6.1 to latest 4.8.2 we started to see period errors like this:
```
System.Transactions.TransactionAbortedException: The transaction has aborted. ---> System.InvalidOperationException: The requested operation cannot be completed because the connection has been broken.

   --- End of inner exception stack trace ---
   at System.Transactions.TransactionStateAborted.EndCommit(InternalTransaction tx)
   at System.Transactions.CommittableTransaction.Commit()
   at System.Transactions.TransactionScope.InternalDispose()
   at System.Transactions.TransactionScope.Dispose()
   ...
```

Those exceptions happening in pure ADO.NET code with TransactionScope created with ReadCommitted isolation level, TransactionScopeOption.Required, 1 minute timeout and TransactionScopeAsyncFlowOption.Enabled.

After rolling back to System.Data.SqlClient 4.8.1 those exceptions dissapeared but we saw another new type of exceptions we've never seen before from cassandra driver:

```
Exception while trying borrow a connection from a pool

System.Net.Sockets.SocketException: Transport endpoint is not connected
   at Cassandra.Connections.HostConnectionPool.GetExistingConnections()
   at Cassandra.Connections.HostConnectionPool.BorrowExistingConnection()
   at async Cassandra.Connections.HostConnectionPool.GetConnectionFromHostAsync(?)
```

Downgrading back to 4.6.1 seems to fixed all issues.

### Configuration

It happened on aspnet core 3.1 web applications and console netcore 3.1 applications with latest 3.1 images both running in container on Linux.

### Regression?

It works on System.Data.SqlClient 4.6.1

### Other information

This issue happening on heavy load application and only on production. I can't reproduce this issue locally.
I wish I could find source for System.Data.SqlClient library to see the diff for 4.8.2 version.

**Comment by MichaelLogutov** [comment]
Oh. I must have misread something, sorry about that. Still, with new 4.8.2 code it could be possible to get no exception from Promote and some other methods if transaction is aborted. I wonder if it's the source of my problems.

You've mentioned to check if transaction is active before performing commit - how do I do that? I have code like this:
```
using (var transaction = TransactionScopeProvider.CreateTransactionScope())
{
    foreach (var x in orderNumbers)
    {
        await this.imagesSqlDataAccess.UpdateAsync(
            new PartialImageUpdateParams
            {
                ImageId = x.ImageId,
                SetOrderNum = x.OrderNumber
            },
            cancellationToken);
    }

    transaction.Complete();
}
```

Where `TransactionScopeProvider.CreateTransactionScope` just a helper to create transaction scope with predefined defaults:
```
var transaction_scope = new TransactionScope(TransactionScopeOption.Required,
    new TransactionOptions
    {
        IsolationLevel = IsolationLevel.ReadCommitted,
        Timeout = TimeSpan.FromMinutes(1)
    },
    TransactionScopeAsyncFlowOption.Enabled);

return transaction_scope;
```

And TransactionScope does not have Status field.
Or you're saying to look into `Transaction.Current.TransactionInformation.Status` ?

**Comment by cheenamalhotra** [comment]
Hi @MichaelLogutov 

> Or you're saying to look into Transaction.Current.TransactionInformation.Status ?

Yes, that's right. Did that work for you?

**Comment by MichaelLogutov** [comment]
I've updated System.Data.SqlClient to 4.8.2 and changed the code to:
```
using (var transaction = TransactionScopeProvider.CreateTransactionScope())
{
    foreach (var x in orderNumbers)
    {
        await this.imagesSqlDataAccess.UpdateAsync(
            new PartialImageUpdateParams
            {
                ImageId = x.ImageId,
                SetOrderNum = x.OrderNumber
            },
            cancellationToken);
    }

    LogCurrentTransactionStatus();

    transaction.Complete();
}

...

private static void LogCurrentTransactionStatus()
{
    try
    {
        var t = Transaction.Current;
        if (t == null)
            "Current transaction is null".Log(LogLevel.Warn);
        else if (t.TransactionInformation.Status != TransactionStatus.Active)
            $"Current transaction status is {t.TransactionInformation.Status}".Log(LogLevel.Warn);
    }
    catch (Exception ex)
    {
        ex.Log("Error while getting current transaction", LogLevel.Warn);
    }
}
```

And I've got exceptions like before:
```
System.Transactions.TransactionException: The operation is not valid for the state of the transaction.
   at System.Transactions.TransactionStatePSPEOperation.get_Status(InternalTransaction tx)
   at System.Transactions.TransactionInformation.get_Status()
   at System.Data.SqlClient.SqlDelegatedTransaction.Promote()
   at System.Transactions.TransactionStatePSPEOperation.PSPEPromote(InternalTransaction tx)
   at System.Transactions.TransactionStateDelegatedBase.EnterState(InternalTransaction tx)
   at System.Transactions.EnlistableStates.Promote(InternalTransaction tx)
   at System.Transactions.Transaction.Promote()
   at System.Transactions.TransactionInterop.ConvertToDistributedTransaction(Transaction transaction)
   at System.Transactions.TransactionInterop.GetExportCookie(Transaction transaction, Byte[] whereabouts)
   at System.Data.SqlClient.SqlInternalConnection.GetTransactionCookie(Transaction transaction, Byte[] whereAbouts)
   at System.Data.SqlClient.SqlInternalConnection.EnlistNonNull(Transaction tx)
   at System.Data.ProviderBase.DbConnectionPool.PrepareConnection(DbConnection owningObject, DbConnectionInternal obj, Transaction transaction)
   at System.Data.ProviderBase.DbConnectionPool.TryGetConnection(DbConnection owningObject, UInt32 waitForMultipleObjectsTimeout, Boolean allowCreate, Boolean onlyOneCheckConnection, DbConnectionOptions userOptions, DbConnectionInternal& connection)
   at System.Data.ProviderBase.DbConnectionPool.TryGetConnection(DbConnection owningObject, TaskCompletionSource`1 retry, DbConnectionOptions userOptions, DbConnectionInternal& connection)
   at System.Data.ProviderBase.DbConnectionFactory.TryGetConnection(DbConnection owningConnection, TaskCompletionSource`1 retry, DbConnectionOptions userOptions, DbConnectionInternal oldConnection, DbConnectionInternal& connection)
   at System.Data.ProviderBase.DbConnectionInternal.TryOpenConnectionInternal(DbConnection outerConnection, DbConnectionFactory connectionFactory, TaskCompletionSource`1 retry, DbConnectionOptions userOptions)
   at System.Data.SqlClient.SqlConnection.TryOpen(TaskCompletionSource`1 retry)
   at System.Data.SqlClient.SqlConnection.OpenAsync(CancellationToken cancellationToken)
   at async Common.Database.SqlDataAccess.Implementation.ImagesSqlDataAccessBase.ExecuteAsync(?) in /app/Common/Database/SqlDataAccess/Implementation/ImagesSqlDataAccessBase.generated.cs:line 297
   at async Common.Services.Implementation.ImagesService.SetOrderNumbersAsync(?) in /app/Common/Services/Implementation/ImagesService.cs:line 135
   at async WebApi.Controllers.EditController.SetOrderNumbersAsync(?) in /app/WebApi/Controllers/EditController.cs:line 47
   at lambda_method(Closure , Object )
   at async Microsoft.AspNetCore.Mvc.Infrastructure.ActionMethodExecutor.AwaitableResultExecutor.Execute(?)
   at async Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.<InvokeActionMethodAsync>g__Awaited|12_0(?)
   at async Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.<InvokeNextActionFilterAsync>g__Awaited|10_0(?)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.Rethrow(ActionExecutedContextSealed context)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.Next(State& next, Scope& scope, Object& state, Boolean& isCompleted)
   at async Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.<InvokeInnerFilterAsync>g__Awaited|13_0(?)
   at async Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.<InvokeFilterPipelineAsync>g__Awaited|19_0(?)
   at async Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.<InvokeAsync>g__Awaited|17_0(?)
   at async Microsoft.AspNetCore.Routing.EndpointMiddleware.<Invoke>g__AwaitRequestTask|6_0(?)
   at async Microsoft.AspNetCore.Authorization.AuthorizationMiddleware.Invoke(?)
   at async Cian.Web.Facilities.ErrorHandling.ErrorHandlingMiddleware.InvokeAsync(?)
```

But still no new warnings about transaction state (there are still other old type of warnings present, so logging is working ok).

There are other examples of the same exception but on different method (if it helps):
```
System.Transactions.TransactionAbortedException: The transaction has aborted. ---> System.InvalidOperationException: The requested operation cannot be completed because the connection has been broken.

   --- End of inner exception stack trace ---
   at System.Transactions.TransactionStateAborted.EndCommit(InternalTransaction tx)
   at System.Transactions.CommittableTransaction.Commit()
   at System.Transactions.TransactionScope.InternalDispose()
   at System.Transactions.TransactionScope.Dispose()
   at async Common.Services.Implementation.ImagesService.AddNewAsync(?) in /app/Common/Services/Implementation/ImagesService.cs:line 295
   at async WebApi.Controllers.AddController.AddAsync(?) in /app/WebApi/Controllers/AddController.cs:line 78
```

**Comment by cheenamalhotra** [comment]
Could you wrap this in a minimal repro to take a closer look at the scenario?

**Comment by MichaelLogutov** [comment]
I would gladly, but I'm afraid I can't - the code is simple ADO.NET update but something could be with the all other stuff microservice doing (heavy CPU usage due to image resizing on Skia + cassandra and rabbitmq libraries connected to servers).

**Comment by cheenamalhotra** [comment]
Ok, I'll try to form a repro based on provided info. 

Could you also confirm if your connection is a pooled connection or pooling is disabled?

Does changing that make any difference?

**Comment by MichaelLogutov** [comment]
> Could you also confirm if your connection is a pooled connection or pooling is disabled?

It's default connection string so it's pooled by default (if I'm not mistaken).
The only thing that changed - max pool size. So connection string looks like this:
`
server=my.server.name; database=MyDatabase; User ID=MyUser; Password=mypassword; Max Pool Size=500
`

> Does changing that make any difference?

Before I tested it, just to clarify - should I just add "Pooling=false" or also change min/max pool size in connection string?

> And also do you also know why the transaction aborts at the first place? Are there failing commands or it's something else? Have you been able to gather it from logs and is it possible to prevent the transaction abort?

I think it doesn't aborts at all. Because otherwise there should be warning about transaction in not active state before commit.

**Comment by cheenamalhotra** [comment]
> should I just add "Pooling=false" or also change min/max pool size in connection string

Just changing "Pooling = false" should disable it.

**Comment by cheenamalhotra** [comment]
> I think it doesn't aborts at all. Because otherwise there should be warning about transaction in not active state before commit.

There's certainly an abort triggered somewhere, otherwise you'd never run into this issue, since the change in last release was made specifically on "Transaction Abort" callback API.

You should be able to capture SQL Profiler traces for "Transaction" > "DTC Transaction" events to see which "Application" triggers "Transaction Aborted" events. You should notice something like below in logs:

| EventClass | TextData | TransactionId | SPID | EventSubClass | DatabaseName |
|:-- |:-- |:-- |:-- |:-- |:-- |
| DTCTransaction | {D3296B44-F55C-4EA2-9B26-3A6011464D6C} | 4502378 | 18 | 16   - Transaction is aborting | master |

**Comment by wolgasster** [comment]
We started getting the same error in .Net Framework 4.8 application under high load (lots of parallel user sessions execute database queries) after upgrade of Microsoft.Data.SqlClient from 1.1.2 to 1.1.3.
Input conditions look the same as in the issue's description. All commands are executed wrapped by a TransactionScope and using connections from ConnectionPool, smth like:

```csharp
using (var transactionScope = DataAccessFactory.Current.CreateTransactionScope())
{
	using (var connection = DataAccessFactory.Current.CreateConnection())
	{
		// create and execute command ...
	}
	transactionScope.Complete();
}
```

**Comment by dotWils** [comment]
Having the same problem here with .NET Framework 4.8 and Microsoft.Data.SqlClient 2.1.1

transaction = new TransactionScope(TransactionScopeOption.RequiresNew, new TransactionOptions()
                {
                    IsolationLevel = System.Transactions.IsolationLevel.ReadUncommitted,
                }, TransactionScopeAsyncFlowOption.Enabled);

**Comment by cheenamalhotra** [comment]
Hi @wolgasster @dotWils 

Is it possible to create a repro application to mimic your scenarios where exception occurs? That would help us in understanding more about your usecase. 

The closing of Connection happens only on Transaction Abort event, and that's important to do because the driver does not lose reference to old transaction and continues to reattach it for next connection. If there's a server side delay to abort old transaction, server accepts that request. This in turn causes a security issue since whenever the transaction does end on the server side, the remaining queries on the new connection execute with implicit transactions.

We may have to look at why an aborted transaction is being committed by your application.

**Comment by dotWils** [comment]
> Hi @wolgasster @dotWils
> 
> Is it possible to create a repro application to mimic your scenarios where exception occurs? That would help us in understanding more about your usecase.
> 
> The closing of Connection happens only on Transaction Abort event, and that's important to do because the driver does not lose reference to old transaction and continues to reattach it for next connection. If there's a server side delay to abort old transaction, server accepts that request. This in turn causes a security issue since whenever the transaction does end on the server side, the remaining queries on the new connection execute with implicit transactions.
> 
> We may have to look at why an aborted transaction is being committed by your application.

Thx for being willing to look at it.
It will be difficult to send a repro as it's a complete engine running procedures on a database.
Our .NET 4.8 application runs for days on .NET 4.8 System.Data.SqlClient. (No transactions aborted)
When only upgrading client to Microsoft.Data.SqlClient (of course in the whole application) we get this rather fast.
Will talk with client if it's worth the effort...

**Comment by csnemes** [comment]
We have the same problem with Microsoft.Data.SqlClient under Net 5.0. With heavy parallel load the connection pooling/transaction handling seems to break down.
With the attached solution the problem can be reproduced. It runs heavy load in cycles. Sometimes you have to wait till cycle 50 or so for the problem to occur, but it will appear consistently.
[SQLProblem.zip](https://github.com/dotnet/SqlClient/files/6272429/SQLProblem.zip)
I tried to dig deeper, but unfortunately the System.Transactions.TransactionEventSource is not working at the moment.

**Comment by cheenamalhotra** [comment]
@csnemes @dotWils @wolgasster @MichaelLogutov 

We have a potential fix for this issue in PR #1042.
I've tested the PR with the repro "SQLproblem" as well from https://github.com/dotnet/SqlClient/issues/729#issuecomment-814978239 and this issue seems to be solved. Would appreciate your feedback if you could test and confirm the fix from this package: [Build Artifacts](https://dev.azure.com/sqlclientdrivers-ci/sqlclient/_build/results?buildId=29567&view=artifacts&pathAsName=false&type=publishedArtifacts)

Thank you

**Comment by csnemes** [comment]
Hi, I've managed to try the patch both with the reproduction and within our application (where we originally encountered with this problem), and it works perfectly. Thanks a lot for the fix.

**Comment by cheenamalhotra** [comment]
Closing this issue as fixed has been merged for main(3.0), 2.1 and 1.1 servicing.
Next releases will contain fix for this issue.

**Comment by dotWils** [comment]
> Closing this issue as fixed has been merged for main(3.0), 2.1 and 1.1 servicing.
> Next releases will contain fix for this issue.

Hi, not so clear how this will be released.
The title reads System.Data.SqlClient will this be version 4.8.3 or will this not be serviced?
Because the merging/servicing you mention is for Microsoft.Data.SqlClient?
